clients/src/main/java/org/apache/kafka/clients/admin/RemoveMemberFromGroupResult.java

```diff
@@ -19,67 +19,53 @@
+ *
+ * The API of this class is evolving, see {@link Admin} for details.
  */
 public class RemoveMemberFromGroupResult {
 
-    private final Errors topLevelError;
-    private final Map<MemberIdentity, KafkaFuture<Void>> memberFutures;
-    private boolean hasError = false;
-
-    RemoveMemberFromGroupResult(LeaveGroupResponse response,
-                                List<MemberIdentity> membersToRemove) {
-        this.topLevelError = response.topLevelError();
-        this.memberFutures = new HashMap<>(membersToRemove.size());
-
-        if (this.topLevelError != Errors.NONE) {
-            // If the populated error is a top-level error, fail every member's future.
-            for (MemberIdentity memberIdentity : membersToRemove) {
-                KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
-                future.completeExceptionally(topLevelError.exception());
-                memberFutures.put(memberIdentity, future);
-            }
-            hasError = true;
-        } else {
-            for (MemberResponse memberResponse : response.memberResponses()) {
-                KafkaFutureImpl<Void> future = new KafkaFutureImpl<>();
-                Errors memberError = Errors.forCode(memberResponse.errorCode());
-                if (memberError != Errors.NONE) {
-                    future.completeExceptionally(memberError.exception());
-                    hasError = true;
-                } else {
-                    future.complete(null);
-                }
-                memberFutures.put(new MemberIdentity()
-                                      .setMemberId(memberResponse.memberId())
-                                      .setGroupInstanceId(memberResponse.groupInstanceId()), future);
-            }
-        }
-    }
+    KafkaFuture<Map<MemberIdentity, Errors>> future;
 
-    public Errors topLevelError() {
-        return topLevelError;
+    RemoveMemberFromGroupResult(KafkaFuture<Map<MemberIdentity, Errors>> future) {
+        this.future = future;
     }
 
-    public boolean hasError() {
-        return hasError;
+    /**
+     * Returns a future which contains the top level error (if any).
+     */
+    public KafkaFuture<Void> all() {
+        return this.future.thenApply(memberResults -> null);
     }
 
     /**
-     * Futures of members with corresponding errors when they leave the group.
-     *
-     * @return list of members who failed to be removed
+     * Returns the selected member future.
      */
-    public Map<MemberIdentity, KafkaFuture<Void>> memberFutures() {
-        return memberFutures;
+    public KafkaFuture<Void> memberResult(MemberIdentity member) {
+        final KafkaFutureImpl<Void> result = new KafkaFutureImpl<>();
+
+        this.future.whenComplete((memberErrors, throwable) -> {
+            if (throwable != null) {
+                result.completeExceptionally(throwable);
+            } else if (!memberErrors.containsKey(member)) {
```

Comment:
@abbccdda @hachikuji I'm wondering if it would ever be a case, that in the response map some entries provided in the options are omitted, because in that case, let's say if all other entries are good, then `all()` would actually succeed while `memberResult` on this member would fail (other `Results` would have similar issues).

Reply:
It's possible but in a user perspective they could still make a wrong query, so I still think an illegal argument exception shouldn't be sth surprising.

Reply:
Yeah but I'm thinking in this case should `all()` throws an exception as well?

Reply:
It's hard to achieve as we only use one level future. If you want to see the actual members who are problematic, `all()` has to be able to complete successfully IMO.

Reply:
The semantics of `all` is defined as "all members are okay, and the global error is okay", that's why I'm wondering if some members are missing from the map, should `all` still be good or not.